osu.Game.Rulesets.Mania/ManiaFilterCriteria.cs

```diff
@@ -17,28 +17,58 @@ namespace osu.Game.Rulesets.Mania
 {
     public class ManiaFilterCriteria : IRulesetFilterCriteria
     {
-        private FilterCriteria.OptionalRange<float> keys;
-
+        private FilterCriteria.OptionalRange<float> included_key_range;
+        private HashSet<int> included_keys = new HashSet<int>();
+        private HashSet<int> excluded_keys = new HashSet<int>();
```

Comment:
All of these should have "key *count*" in the name too.

Also, because key count is functionally limited to 1..20 range, you can replace all three of these with a hash set that's default-initialized to 1..20 and then remove from it as required.